staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go

```diff
@@ -85,79 +76,56 @@ func NewMutatingWebhook(configFile io.Reader) (*MutatingWebhook, error) {
-			admission.Connect,
-			admission.Create,
-			admission.Delete,
-			admission.Update,
-		),
+	return &Webhook{
+		Handler:       handler,
+		sourceFactory: sourceFactory,
 		clientManager: cm,
+		dispatcher:    dispatcherFactory(&cm),
 	}, nil
 }
 
-var _ admission.MutationInterface = &MutatingWebhook{}
-
-// MutatingWebhook is an implementation of admission.Interface.
-type MutatingWebhook struct {
-	*admission.Handler
-	hookSource       generic.Source
-	namespaceMatcher namespace.Matcher
-	clientManager    config.ClientManager
-	convertor        versioned.Convertor
-	jsonSerializer   runtime.Serializer
-}
-
-var (
-	_ = genericadmissioninit.WantsExternalKubeClientSet(&MutatingWebhook{})
-)
-
 // TODO find a better way wire this, but keep this pull small for now.
-func (a *MutatingWebhook) SetAuthenticationInfoResolverWrapper(wrapper config.AuthenticationInfoResolverWrapper) {
+func (a *Webhook) SetAuthenticationInfoResolverWrapper(wrapper config.AuthenticationInfoResolverWrapper) {
 	a.clientManager.SetAuthenticationInfoResolverWrapper(wrapper)
 }
 
 // SetServiceResolver sets a service resolver for the webhook admission plugin.
 // Passing a nil resolver does not have an effect, instead a default one will be used.
-func (a *MutatingWebhook) SetServiceResolver(sr config.ServiceResolver) {
+func (a *Webhook) SetServiceResolver(sr config.ServiceResolver) {
 	a.clientManager.SetServiceResolver(sr)
 }
 
 // SetScheme sets a serializer(NegotiatedSerializer) which is derived from the scheme
-func (a *MutatingWebhook) SetScheme(scheme *runtime.Scheme) {
+func (a *Webhook) SetScheme(scheme *runtime.Scheme) {
 	if scheme != nil {
 		a.clientManager.SetNegotiatedSerializer(serializer.NegotiatedSerializerWrapper(runtime.SerializerInfo{
 			Serializer: serializer.NewCodecFactory(scheme).LegacyCodec(admissionv1beta1.SchemeGroupVersion),
 		}))
 		a.convertor.Scheme = scheme
-		a.jsonSerializer = json.NewSerializer(json.DefaultMetaFactory, scheme, scheme, false)
 	}
 }
 
 // WantsExternalKubeClientSet defines a function which sets external ClientSet for admission plugins that need it
-func (a *MutatingWebhook) SetExternalKubeClientSet(client clientset.Interface) {
+func (a *Webhook) SetExternalKubeClientSet(client clientset.Interface) {
 	a.namespaceMatcher.Client = client
 }
 
 // SetExternalKubeInformerFactory implements the WantsExternalKubeInformerFactory interface.
-func (a *MutatingWebhook) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+func (a *Webhook) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
 	namespaceInformer := f.Core().V1().Namespaces()
 	a.namespaceMatcher.NamespaceLister = namespaceInformer.Lister()
 	mutatingWebhookConfigurationsInformer := f.Admissionregistration().V1beta1().MutatingWebhookConfigurations()
-	a.hookSource = configuration.NewMutatingWebhookConfigurationManager(mutatingWebhookConfigurationsInformer)
+	a.hookSource = a.sourceFactory(f)
 	a.SetReadyFunc(func() bool {
```

Comment:
same as above, if we SetReadyFunc here, I think `https://github.com/sttts/kubernetes/blob/18e7408eb225014cd25d25f57f37ef3a8165d245/staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go#L47` does not need.

Reply:
fixed